### ISSUE TITLE
Copy to clipboard inserts editor styling and fonts

### ISSUE THREAD

**Issue by glassdimly** [atom-community/markdown-preview-plus#418]
When I copy from the preview pane to the clipboard using "Copy as HTML" I would expect sparse output, with lean HTML. Instead, this is what I get:

```
<p style="box-sizing: border-box; margin: 0px 0px 15px; line-height: 1.6; color: rgb(60, 173, 215); font-family: ".SFNSText-Regular", "SF UI Text", "Lucida Grande", "Segoe UI", Ubuntu, Cantarell, sans-serif; font-size: 14.4px; font-variant-ligatures: normal; orphans: 2; widows: 2; background-color: rgb(14, 21, 27);">
My content here. <em style="box-sizing: border-box;">Emphasized.</em>
</p>
```
This is the output I would expect:

```
<p>
My content here. <em>Emphasized.</em>
</p>
```

Could there be an option for this? It seems that most people wouldn't want all that formatting, but some might.

I plan to have a crack at solving this. Is the output of styles a bug, or expected behavior? Is the addition of an option an acceptable solution, or not?

**Comment by glassdimly** [comment]
From `markdown-preview-plus.cson` it looks like you're using `core:copy`:

`{label: 'Copy As HTML', command: 'core:copy'}`

**Comment by lierdakil** [comment]
Don't let `core:copy` mislead you, it's overloaded and will behave differently under different circumstances.

So, when you have an active selection in the preview, 'Copy As HTML' will use Atom's default handler, mostly because there's no sure-fire way to reconstruct the initial Markdown. And we'd need initial markdown to get "lean HTML" out, because copying displayed HTML source verbatim, for one, isn't straightforward, and it isn't what most users would expect here anyway (because in many cases, math being one of those, copied HTML won't make any sense). The default Atom handler works more or less as Chrome does, i.e. copies the rich text content, which you could then paste somewhere appropriate. It works, but as you noted, it carries *way* more styles than strictly necessary.

When you *don't* have an active selection, that's when the overload kicks in. In that case, source Markdown is re-rendered for export in the background, and the result gets put into the clipboard. That's your "lean HTML". The caveat is: it's always the whole document.

You can, however, "copy-as-html" parts of a Markdown document. `markdown-preview-plus:copy-html` will do exactly that, but it works on editor, not preview.

So, to summarize, there is a way to do something similar to what you want, that is, copy part of your Markdown as lean HTML. But you need to select the *source* snippet you want to copy as HTML, not a part of the preview, and run a special command. Bear in mind this is not foolproof and might produce unexpected results with complicated markups -- feel free to file issues if you run into something like that.

Copying parts of the preview is ultimately handled by Chromium, so you can blame it for the behaviour you're describing.

**Comment by glassdimly** [comment]
Thanks for your detailed reply! 

I believe I haven't communicated properly here. I don't want to copy raw HTML, but rather _HTML encoded for the clipboard._

* `markdown-preview-plus:copy-html` copies raw HTML source
* `Copy Current File to the Clipbard` copies as raw HTML source

My workflow is this: I compose an email in MD; I then want to copy that MD as HTML to the clipboard to paste directly into gmail (for example), and I don't want a bunch of styles to transfer over.

How best can I accomplish this? I'm not averse to writing my own plugin or contributing.

**Comment by glassdimly** [comment]
It might not be too hard to overload `copyHtml(text, filePath, renderLaTeX)` so that instead of `atom.clipboard.write(res)` it uses [this electron clipboard method](https://github.com/electron/electron/blob/master/docs/api/clipboard.md#clipboardwritehtmlmarkup-type). 

Does that seem reasonable? Bear in mind I don't know what objects, etc, are available yet as I'm still learning about Atom app dev so let me know if my assumptions are wrong. I.e. is `atom.clipboard` the electron clipboard... or maybe it's overloaded and sometimes calls that clipboard..?

**Comment by lierdakil** [comment]
Calling Electron's clipboard directly is an option. Although I would insist that `writeHTML` itself is bad because it doesn't provide a plaintext fallback, but `clipboard.write` would work. I could slap together a prototype in a bit I believe.

**Comment by lierdakil** [comment]
See #419. If you can test it and report if it works for your purposes, that would be nice. If not, from my (admittedly limited) testing it works more or less as I would expect: pastes rich text into rich editors and plain raw HTML in plain editors. This behaviour can be disabled with a config option.

Some tests fail, so this still needs some work, but the basic idea seems workable.
![mpp-rich-clipboard](https://user-images.githubusercontent.com/7275622/44279897-d1c6f200-a25b-11e8-800f-bde880d7e9ac.gif)

**Comment by glassdimly** [comment]
I will test. I've left a comment or two on the PR.

**Comment by glassdimly** [comment]
...and thank you!

**Comment by lierdakil** [comment]
Released as v3.4.0. Feel free to open new issues if something's amiss.

**Comment by glassdimly** [comment]
I tested, and it works great for what I'm doing.

I think it would be nice to have a context menu for this so I made a PR [here](https://github.com/atom-community/markdown-preview-plus/pull/420).
